pkg/pgmodel/ingestor/watcher.go

```diff
@@ -0,0 +1,107 @@
+	reportRatioThreshold = 3
+	refreshIteration     = 10 // This means that values of watcher.incomingBatches and watcher.outgoingBatches gets renewed every (refreshIteration * checkThroughputInterval).
+	checkRatioInterval   = time.Minute
+
+	// Duration warnings constants.
+	reportDurationThreshold        = time.Minute
+	checkIngestionDurationInterval = time.Minute
+)
+
+// throughtputWatcher is a light weight samples batch watching type, that serves as a watching
+// routine and keeps the track of incoming write batches. It keeps a ratio of
+// incoming samples vs outgoing samples and warn about low throughput to the user
+// which might be due to external causes like network latency, resources allocated, etc.
+//
+// We watch the batches than the samples in total, since watching batching is fast
+// and does the same purpose of watching samples, but on a higher-level.
+type throughtputWatcher struct {
+	// Warn based on ratio.
+	incomingBatches uint64
+	outgoingBatches uint64
+
+	// Warn based on ingestion duration.
+	shouldReportLastDuration atomic.Value
+	lastIngestionDuration    atomic.Value
+}
+
+var watcher *throughtputWatcher
+
+func init() {
+	watcher = &throughtputWatcher{
+		incomingBatches: 0,
+		outgoingBatches: 0,
+	}
+	watcher.shouldReportLastDuration.Store(false)
+	go watcher.watch()
+	go watcher.watchTime()
+}
+
+// watch watches the ratio between incomingBatches and outgoingBatches every checkThroughputInterval.
+func (w *throughtputWatcher) watch() {
+	var (
+		itr uint8
+		t   = time.NewTicker(checkRatioInterval)
+	)
+	for range t.C {
+		if w.outgoingBatches > 1 {
+			if ratio := float64(atomic.LoadUint64(&w.incomingBatches)) / float64(atomic.LoadUint64(&w.outgoingBatches)); ratio > reportRatioThreshold {
+				w.warn(ratio)
+			}
+		}
+		if itr > refreshIteration {
+			atomic.StoreUint64(&watcher.incomingBatches, 0)
+			atomic.StoreUint64(&watcher.outgoingBatches, 0)
+		}
+	}
+}
+
+func (w *throughtputWatcher) warn(ratio float64) {
+	log.Warn("msg", "[WARNING] Incoming samples rate higher than outgoing samples. This may happen due to poor network latency, "+
+		"less resources allocated, or few concurrent writes (shards) from Prometheus. Please tune your resource to improve performance. ",
+		"Throughput ratio", fmt.Sprintf("%.2f", ratio), "Ratio threshold", reportRatioThreshold)
+}
+
+func (w *throughtputWatcher) watchTime() {
+	t := time.NewTicker(checkIngestionDurationInterval)
+	for range t.C {
+		if w.shouldReportLastDuration.Load().(bool) {
+			w.shouldReportLastDuration.Store(false)
+			log.Warn("msb", "[WARNING] Samples ingestion duration too high: Please make sure you have low network latency and "+
+				"proper resources allocated.",
+				"Last ingestion duration", w.lastIngestionDuration.Load().(time.Duration).String())
+		}
+	}
+}
+
+func incSamples(size uint64) {
+	atomic.AddUint64(&watcher.incomingBatches, size)
+}
+
+func decSamples(size uint64) {
+	atomic.AddUint64(&watcher.outgoingBatches, size)
+}
```

Comment:
can we use a prometheus gauge? I think we can handle an approximate value for this number no need to go atomic